### PR TITLE
FEAT: add topic subscription option to message pump

### PR DESCRIPTION
**Pull request by stijnmoreels**
Add an option to include the creation/deletion Auzre Service Bus Topic subscription during the start/stop of the message pump.

#39

### CODE REVIEW

**Comment by stijnmoreels** [comment]
Was it something like this you had in mind?

**Comment by tomkerkhove** [comment]
Had a quick look and approach looks good, but will provide better thorough review soon.

Would you mind updating the docs as well please?

File: src/Arcus.Messaging.Pumps.ServiceBus/AzureServiceBusMessagePump.cs [hunk 1]
```diff
@@ -35,7 +37,9 @@ public abstract class AzureServiceBusMessagePump<TMessage> : MessagePump<TMessag
             : base(configuration, serviceProvider, logger)
         {
             Settings = serviceProvider.GetRequiredService<AzureServiceBusMessagePumpSettings>();
+            JobId = Settings.Options.JobId;
 
+            _subscriptionName = $"{Settings.SubscriptionPrefix}:{JobId}";
```
Comment (tomkerkhove) [inline src/Arcus.Messaging.Pumps.ServiceBus/AzureServiceBusMessagePump.cs]: Let's use a hyphen instead, this is more commonly used
```suggestion
            _subscriptionName = $"{Settings.SubscriptionPrefix}-{JobId}";
```

File: src/Arcus.Messaging.Pumps.ServiceBus/AzureServiceBusMessagePump.cs [hunk 2]
```diff
@@ -49,6 +53,42 @@ public abstract class AzureServiceBusMessagePump<TMessage> : MessagePump<TMessag
...
+
+        /// <summary>
+        /// Triggered when the application host is ready to start the service.
+        /// </summary>
+        /// <param name="cancellationToken">Indicates that the start process has been aborted.</param>
+        public override async Task StartAsync(CancellationToken cancellationToken)
+        {
+            if (Settings.Options.IncludeTopicSubscription)
+            {
+                ServiceBusConnectionStringBuilder serviceBusConnectionString = await GetServiceBusConnectionStringAsync();
```
Comment (tomkerkhove) [inline src/Arcus.Messaging.Pumps.ServiceBus/AzureServiceBusMessagePump.cs]: Let's move all of this to a dedicated method

File: src/Arcus.Messaging.Pumps.ServiceBus/AzureServiceBusMessagePump.cs [hunk 3]
```diff
@@ -49,6 +53,42 @@ public abstract class AzureServiceBusMessagePump<TMessage> : MessagePump<TMessag
...
+
+                Logger.LogTrace("[Job: {JobId}] Creating subscription '{SubscriptionName}' on topic '{TopicPath}'...", JobId, _subscriptionName, serviceBusConnectionString.EntityPath);
```
Comment (tomkerkhove) [inline src/Arcus.Messaging.Pumps.ServiceBus/AzureServiceBusMessagePump.cs]: Do we use the `[Job: {JobId}] ` prefix for other traces as well? If not, please align this line or the others

File: src/Arcus.Messaging.Pumps.ServiceBus/AzureServiceBusMessagePump.cs [hunk 4]
```diff
@@ -49,6 +53,42 @@ public abstract class AzureServiceBusMessagePump<TMessage> : MessagePump<TMessag
...
+                var subscriptionDescription = new SubscriptionDescription(serviceBusConnectionString.EntityPath, _subscriptionName)
+                {
+                    AutoDeleteOnIdle = TimeSpan.FromHours(1),
+                    MaxDeliveryCount = 3,
+                    UserMetadata = $"Subscription created by Arcus job: '{JobId}' to process inbound CloudEvents."
```
Comment (tomkerkhove) [inline src/Arcus.Messaging.Pumps.ServiceBus/AzureServiceBusMessagePump.cs]: Copy paste?

File: src/Arcus.Messaging.Pumps.ServiceBus/AzureServiceBusMessagePump.cs [hunk 5]
```diff
@@ -49,6 +53,42 @@ public abstract class AzureServiceBusMessagePump<TMessage> : MessagePump<TMessag
...
+        public override async Task StartAsync(CancellationToken cancellationToken)
+        {
+            if (Settings.Options.IncludeTopicSubscription)
+            {
+                ServiceBusConnectionStringBuilder serviceBusConnectionString = await GetServiceBusConnectionStringAsync();
+
+                Logger.LogTrace("[Job: {JobId}] Creating subscription '{SubscriptionName}' on topic '{TopicPath}'...", JobId, _subscriptionName, serviceBusConnectionString.EntityPath);
+                var subscriptionDescription = new SubscriptionDescription(serviceBusConnectionString.EntityPath, _subscriptionName)
+                {
+                    AutoDeleteOnIdle = TimeSpan.FromHours(1),
```
Comment (tomkerkhove) [inline src/Arcus.Messaging.Pumps.ServiceBus/AzureServiceBusMessagePump.cs]: Let's not set this one so that ops can troubleshoot things

File: src/Arcus.Messaging.Pumps.ServiceBus/AzureServiceBusMessagePump.cs [hunk 6]
```diff
@@ -49,6 +53,42 @@ public abstract class AzureServiceBusMessagePump<TMessage> : MessagePump<TMessag
...
+                    MaxDeliveryCount = 3,
```
Comment (tomkerkhove) [inline src/Arcus.Messaging.Pumps.ServiceBus/AzureServiceBusMessagePump.cs]: Not really up to us to define so I'd suggest to leave this out for now

File: src/Arcus.Messaging.Pumps.ServiceBus/AzureServiceBusMessagePump.cs [hunk 7]
```diff
@@ -49,6 +53,42 @@ public abstract class AzureServiceBusMessagePump<TMessage> : MessagePump<TMessag
...
+                    UserMetadata = $"Subscription created by Arcus job: '{JobId}' to process inbound CloudEvents."
+                };
+            
+                var ruleDescription = new RuleDescription("Accept-All", new TrueFilter());
+
+                var serviceBusClient = new ManagementClient(serviceBusConnectionString);
+                await serviceBusClient.CreateSubscriptionAsync(subscriptionDescription, ruleDescription, cancellationToken)
+                                      .ConfigureAwait(continueOnCapturedContext: false);
```
Comment (tomkerkhove) [inline src/Arcus.Messaging.Pumps.ServiceBus/AzureServiceBusMessagePump.cs]: We don't do this in other places nor it's a library, maybe best to remove for now then or not?

File: src/Arcus.Messaging.Pumps.ServiceBus/AzureServiceBusMessagePump.cs [hunk 8]
```diff
@@ -49,6 +53,42 @@ public abstract class AzureServiceBusMessagePump<TMessage> : MessagePump<TMessag
...
+
+                Logger.LogTrace("[Job: {JobId}] Subscription '{SubscriptionName}' created on topic '{TopicPath}'", JobId, _subscriptionName, serviceBusConnectionString.EntityPath);
+                await serviceBusClient.CloseAsync().ConfigureAwait(continueOnCapturedContext: false);
```
Comment (tomkerkhove) [inline src/Arcus.Messaging.Pumps.ServiceBus/AzureServiceBusMessagePump.cs]: We don't do this in other places nor it's a library, maybe best to remove for now then or not?

File: src/Arcus.Messaging.Pumps.ServiceBus/AzureServiceBusMessagePump.cs [hunk 9]
```diff
@@ -49,6 +53,42 @@ public abstract class AzureServiceBusMessagePump<TMessage> : MessagePump<TMessag
...
+                    UserMetadata = $"Subscription created by Arcus job: '{JobId}' to process inbound CloudEvents."
+                };
+            
+                var ruleDescription = new RuleDescription("Accept-All", new TrueFilter());
+
+                var serviceBusClient = new ManagementClient(serviceBusConnectionString);
+                await serviceBusClient.CreateSubscriptionAsync(subscriptionDescription, ruleDescription, cancellationToken)
+                                      .ConfigureAwait(continueOnCapturedContext: false);
+
+                Logger.LogTrace("[Job: {JobId}] Subscription '{SubscriptionName}' created on topic '{TopicPath}'", JobId, _subscriptionName, serviceBusConnectionString.EntityPath);
```
Comment (tomkerkhove) [inline src/Arcus.Messaging.Pumps.ServiceBus/AzureServiceBusMessagePump.cs]: Same here about aligning

File: src/Arcus.Messaging.Pumps.ServiceBus/AzureServiceBusMessagePump.cs (hunk 1 above)
Comment (tomkerkhove) [inline src/Arcus.Messaging.Pumps.ServiceBus/AzureServiceBusMessagePump.cs]: Let's make it a property instead

File: src/Arcus.Messaging.Pumps.ServiceBus/AzureServiceBusMessagePump.cs [hunk 10]
```diff
@@ -250,10 +290,31 @@ protected virtual IEnumerable<ServiceBusPlugin> DefineServiceBusPlugins()
         /// <param name="cancellationToken">Indicates that the shutdown process should no longer be graceful.</param>
         public override async Task StopAsync(CancellationToken cancellationToken)
         {
+            if (Settings.Options.IncludeTopicSubscription)
```
Comment (tomkerkhove) [inline src/Arcus.Messaging.Pumps.ServiceBus/AzureServiceBusMessagePump.cs]: Creating is one thing, deleting is another.

I'd suggest to leave this out or do this via another config option.

File: src/Arcus.Messaging.Pumps.ServiceBus/AzureServiceBusMessagePumpOptions.cs [hunk 11]
```diff
@@ -34,9 +34,28 @@ public class AzureServiceBusMessagePumpOptions
...
 
+        /// <summary>
+        /// Gets or sets the unique identifier for this background job to distinguish this job instance in a multi-instance deployment.
+        /// </summary>
+        public string JobId { get; set; }
+
+        /// <summary>
+        /// Gets or sets the value indicating whether or not a new Azure Service Bus Topic subscription has to be created when the <see cref="AzureServiceBusMessagePump{TMessage}"/> starts.
+        /// The subscription will be deleted afterwards when the message pump stops.
+        /// REMARK: requires at least 'Manage' permissions on the Azure Service Bus Topic to create and delete these subscriptions.
```
Comment (tomkerkhove) [inline src/Arcus.Messaging.Pumps.ServiceBus/AzureServiceBusMessagePumpOptions.cs]: Can be removed since it's in the `remarks` section
```suggestion
```

File: src/Arcus.Messaging.Pumps.ServiceBus/AzureServiceBusMessagePumpOptions.cs [hunk 12]
```diff
@@ -34,9 +34,28 @@ public class AzureServiceBusMessagePumpOptions
...
+        /// </summary>
+        /// <remarks>
+        ///     Requires at least 'Manage' permissions on the Azure Service Bus Topic to create and delete these subscriptions.
```
Comment (tomkerkhove) [inline src/Arcus.Messaging.Pumps.ServiceBus/AzureServiceBusMessagePumpOptions.cs]: Manage is the highest permission
```suggestion
        ///     Requires 'Manage' permissions on the Azure Service Bus Topic to create and delete these subscriptions.
```

File: src/Arcus.Messaging.Pumps.ServiceBus/AzureServiceBusMessagePumpOptions.cs [hunk 13]
```diff
@@ -34,9 +34,28 @@ public class AzureServiceBusMessagePumpOptions
...
         public bool AutoComplete { get; set; }
 
+        /// <summary>
+        /// Gets or sets the unique identifier for this background job to distinguish this job instance in a multi-instance deployment.
+        /// </summary>
+        public string JobId { get; set; }
+
+        /// <summary>
+        /// Gets or sets the value indicating whether or not a new Azure Service Bus Topic subscription has to be created when the <see cref="AzureServiceBusMessagePump{TMessage}"/> starts.
+        /// The subscription will be deleted afterwards when the message pump stops.
```
Comment (tomkerkhove) [inline src/Arcus.Messaging.Pumps.ServiceBus/AzureServiceBusMessagePumpOptions.cs]: See other comment - I'd do thise via seperate option or not at all

File: src/Arcus.Messaging.Pumps.ServiceBus/AzureServiceBusMessagePumpSettings.cs [hunk 14]
```diff
@@ -22,12 +22,12 @@ public class AzureServiceBusMessagePumpSettings
         ///     Constructor
         /// </summary>
         /// <param name="entityName">Name of the entity to process</param>
-        /// <param name="subscriptionName">Name of the subscription to process</param>
+        /// <param name="subscriptionPrefix">Name of the subscription to process</param>
```
Comment (tomkerkhove) [inline src/Arcus.Messaging.Pumps.ServiceBus/AzureServiceBusMessagePumpSettings.cs]:
```suggestion
        /// <param name="subscriptionPrefix">Name prefix of the subscription to process</param>
```

File: src/Arcus.Messaging.Pumps.ServiceBus/AzureServiceBusMessagePumpOptions.cs [hunk 15]
```diff
@@ -34,9 +34,28 @@ public class AzureServiceBusMessagePumpOptions
...
+        /// REMARK: requires at least 'Manage' permissions on the Azure Service Bus Topic to create and delete these subscriptions.
+        /// </summary>
+        /// <remarks>
+        ///     Requires at least 'Manage' permissions on the Azure Service Bus Topic to create and delete these subscriptions.
+        /// </remarks>
+        public bool IncludeTopicSubscription { get; set; }
```
[the remark this answers is not in the document]
Comment (stijnmoreels) [reply src/Arcus.Messaging.Pumps.ServiceBus/AzureServiceBusMessagePumpOptions.cs]: Yeah, but I also wanted to have in the name something that would 'destroy' the subscription too, you know?

File: src/Arcus.Messaging.Pumps.ServiceBus/AzureServiceBusMessagePump.cs (hunk 2 above)
Comment (stijnmoreels) [reply src/Arcus.Messaging.Pumps.ServiceBus/AzureServiceBusMessagePump.cs]: Yes.

File: src/Arcus.Messaging.Pumps.ServiceBus/AzureServiceBusMessagePump.cs (hunk 5 above)
Comment (stijnmoreels) [reply src/Arcus.Messaging.Pumps.ServiceBus/AzureServiceBusMessagePump.cs]: Yes.

File: src/Arcus.Messaging.Pumps.ServiceBus/AzureServiceBusMessagePump.cs (hunk 4 above)
Comment (stijnmoreels) [reply src/Arcus.Messaging.Pumps.ServiceBus/AzureServiceBusMessagePump.cs]: Woops.

File: src/Arcus.Messaging.Pumps.ServiceBus/AzureServiceBusMessagePump.cs (hunk 6 above)
Comment (stijnmoreels) [reply src/Arcus.Messaging.Pumps.ServiceBus/AzureServiceBusMessagePump.cs]: Ok.

File: src/Arcus.Messaging.Pumps.ServiceBus/AzureServiceBusMessagePump.cs (hunk 7 above)
Comment (stijnmoreels) [reply src/Arcus.Messaging.Pumps.ServiceBus/AzureServiceBusMessagePump.cs]: I guess, the creation/deletion of topic subscription can be handled independent of the actual message pump; that's why I've left it here.

File: src/Arcus.Messaging.Pumps.ServiceBus/AzureServiceBusMessagePump.cs (hunk 1 above)
Comment (stijnmoreels) [reply src/Arcus.Messaging.Pumps.ServiceBus/AzureServiceBusMessagePump.cs]: Ok.

File: src/Arcus.Messaging.Pumps.ServiceBus/AzureServiceBusMessagePump.cs (hunk 3 above)
Comment (stijnmoreels) [reply src/Arcus.Messaging.Pumps.ServiceBus/AzureServiceBusMessagePump.cs]: For all log traces, yes, but not all log messages. Like information messages doesn't have this. Do you want me to align it to every log entry that's being written in the pump?

Comment (tomkerkhove) [reply src/Arcus.Messaging.Pumps.ServiceBus/AzureServiceBusMessagePump.cs]: Let's do that, but in a follow-up PR

File: src/Arcus.Messaging.Pumps.ServiceBus/AzureServiceBusMessagePumpOptions.cs (hunk 15 above)
[the remark this answers is not in the document]
Comment (stijnmoreels) [reply src/Arcus.Messaging.Pumps.ServiceBus/AzureServiceBusMessagePumpOptions.cs]: Yep, updated with an enum instead.

File: src/Arcus.Messaging.Pumps.ServiceBus/AzureServiceBusMessagePump.cs (hunk 3 above)
Comment (stijnmoreels) [reply src/Arcus.Messaging.Pumps.ServiceBus/AzureServiceBusMessagePump.cs]: Ref: https://github.com/arcus-azure/arcus.messaging/issues/58

File: src/Arcus.Messaging.Pumps.ServiceBus/AzureServiceBusMessagePumpOptions.cs [hunk 16]
```diff
@@ -46,15 +46,16 @@ public class AzureServiceBusMessagePumpOptions
...
 
         /// <summary>
         ///     Default settings
         /// </summary>
         internal static AzureServiceBusMessagePumpOptions Default => new AzureServiceBusMessagePumpOptions
         {
             AutoComplete = true,
-            JobId = Guid.NewGuid().ToString()
+            JobId = Guid.NewGuid().ToString(),
+            TopicSubscription = TopicSubscription.None
```
Comment (tomkerkhove) [inline src/Arcus.Messaging.Pumps.ServiceBus/AzureServiceBusMessagePumpOptions.cs]: I would turn on subscription creation by default

File: src/Arcus.Messaging.Pumps.ServiceBus/AzureServiceBusMessagePumpOptions.cs [hunk 17]
```diff
@@ -46,15 +46,16 @@ public class AzureServiceBusMessagePumpOptions
         /// <remarks>
         ///     Requires 'Manage' permissions on the Azure Service Bus Topic to create and delete these subscriptions.
```
Comment (tomkerkhove) [inline src/Arcus.Messaging.Pumps.ServiceBus/AzureServiceBusMessagePumpOptions.cs]: How about this?
```suggestion
        ///     Provides capability to create and delete these subscriptions. This requires 'Manage' permissions on the Azure Service Bus Topic or namespace.
```

File: src/Arcus.Messaging.Pumps.ServiceBus/AzureServiceBusMessagePump.cs [hunk 18]
```diff
@@ -49,6 +52,55 @@ public abstract class AzureServiceBusMessagePump<TMessage> : MessagePump<TMessag
...
+        {
+            if ((Settings.Options.TopicSubscription & TopicSubscription.CreateOnStart) == TopicSubscription.CreateOnStart)
+            {
+                await CreateTopicSubscriptionAsync(cancellationToken);
+            }
+
+            await base.StartAsync(cancellationToken);
+        }
+
+        private async Task CreateTopicSubscriptionAsync(CancellationToken cancellationToken)
```
Comment (tomkerkhove) [inline src/Arcus.Messaging.Pumps.ServiceBus/AzureServiceBusMessagePump.cs]: Thinking out loud - We could move this to a `AzureServiceBusTopicManager` or so, but not a must.

File: src/Arcus.Messaging.Pumps.ServiceBus/AzureServiceBusMessagePump.cs [hunk 19]
```diff
@@ -250,10 +302,43 @@ protected virtual IEnumerable<ServiceBusPlugin> DefineServiceBusPlugins()
...
+            if ((Settings.Options.TopicSubscription & TopicSubscription.DeleteOnStop) == TopicSubscription.DeleteOnStop)
+            {
+                await DeleteTopicSubscriptionAsync(cancellationToken);
+            }
+
             await base.StopAsync(cancellationToken);
             _isHostShuttingDown = true;
         }
 
+        private async Task DeleteTopicSubscriptionAsync(CancellationToken cancellationToken)
```
Comment (tomkerkhove) [inline src/Arcus.Messaging.Pumps.ServiceBus/AzureServiceBusMessagePump.cs]: Thinking out loud - We could move this to a `AzureServiceBusTopicManager` or so, but not a must.

Comment (tomkerkhove) [reply src/Arcus.Messaging.Pumps.ServiceBus/AzureServiceBusMessagePump.cs]: If so, a follow-up PR for this might be better maybe 🤔

File: src/Arcus.Messaging.Pumps.ServiceBus/AzureServiceBusMessagePump.cs [hunk 20]
```diff
@@ -250,10 +302,43 @@ protected virtual IEnumerable<ServiceBusPlugin> DefineServiceBusPlugins()
         /// <param name="cancellationToken">Indicates that the shutdown process should no longer be graceful.</param>
         public override async Task StopAsync(CancellationToken cancellationToken)
         {
+            if ((Settings.Options.TopicSubscription & TopicSubscription.DeleteOnStop) == TopicSubscription.DeleteOnStop)
```
Comment (tomkerkhove) [inline src/Arcus.Messaging.Pumps.ServiceBus/AzureServiceBusMessagePump.cs]: Is this a shorthand notation of `Settings.Options.TopicSubscription != null && Settings.Options.TopicSubscription == TopicSubscription.DeleteOnStop`?

**Comment by stijnmoreels** [comment]
I actually don't understand why the integration tests fails. It looks like a configuration issue rather than implementation. :(

File: src/Arcus.Messaging.Pumps.ServiceBus/AzureServiceBusMessagePump.cs (hunk 20 above)
Comment (stijnmoreels) [reply src/Arcus.Messaging.Pumps.ServiceBus/AzureServiceBusMessagePump.cs]: The enum is actually just a number (so can't be `null`). I added the `[Flags]` attribute to indicate that we expect combinations of the enum values.

So, the notation is a bitwise condition to filter out first the `.DeleteOnStop` value (which is `3` from the configured options value (`1` for example).
Bitwise is `3`, `0011` and `1`, `0001`. The `&` filters out all other bits. Like if we would have set the options to `7` (`0111`) it would become `0011`.
This final value (`0011`) we can use to match against the thing we actually want to check `3`.

So, in short, the syntax allows you to first filter out ("clean" if you will) the bits that would make up the expected value, and then check if it's indeed the expected value.
You follow?

File: src/Arcus.Messaging.Pumps.ServiceBus/AzureServiceBusMessagePump.cs (hunk 19 above)
Comment (stijnmoreels) [reply src/Arcus.Messaging.Pumps.ServiceBus/AzureServiceBusMessagePump.cs]: If it would delete the topic on dispose, I would make it a `IDisposable`; but that's not the case. So, hmm, maybe just too little functionality to get the added value to move this?

**Comment by stijnmoreels** [comment]
> Docs will be via a dedicated PR I presume or think it's not worth it?

I can add the docs here. But yeah, there's not much to document 😄 .

File: src/Arcus.Messaging.Pumps.ServiceBus/AzureServiceBusMessagePumpOptions.cs (hunk 16 above)
Comment (stijnmoreels) [reply src/Arcus.Messaging.Pumps.ServiceBus/AzureServiceBusMessagePumpOptions.cs]: Creation but not deletion?

File: src/Arcus.Messaging.Pumps.ServiceBus/AzureServiceBusMessagePump.cs (hunk 19 above)
Comment (tomkerkhove) [reply src/Arcus.Messaging.Pumps.ServiceBus/AzureServiceBusMessagePump.cs]: Would be seperation of concerns but it's ok

File: src/Arcus.Messaging.Pumps.ServiceBus/AzureServiceBusMessagePump.cs (hunk 20 above)
Comment (tomkerkhove) [reply src/Arcus.Messaging.Pumps.ServiceBus/AzureServiceBusMessagePump.cs]: Can't we just use `Settings.Options.TopicSubscription.HasFlag(TopicSubscription.DeleteOnStop)` which does what you want to achieve?

**Comment by tomkerkhove** [comment]
> I actually don't understand why the integration tests fails. It looks like a configuration issue rather than implementation. :(

I'll have a look, maybe it's because it doesn't have management permissions; but if that is the case it would be ideal to provide good logs around that (maybe local testing is better to figure this out though)

File: src/Arcus.Messaging.Pumps.ServiceBus/AzureServiceBusMessagePump.cs (hunk 20 above)
Comment (stijnmoreels) [reply src/Arcus.Messaging.Pumps.ServiceBus/AzureServiceBusMessagePump.cs]: Ah, yes, that would make the code maybe more expressive, but the really downside on that approach is that it has a unnecessary boxing/unboxing on the integer bc it also checks the type of the enum. Bitwise operations are much faster here.
So, ... IDK. I can change it if you like. There's not that much to win I guess.

File: src/Arcus.Messaging.Pumps.ServiceBus/AzureServiceBusMessagePumpOptions.cs (hunk 16 above)
Comment (tomkerkhove) [reply src/Arcus.Messaging.Pumps.ServiceBus/AzureServiceBusMessagePumpOptions.cs]: Jup

File: src/Arcus.Messaging.Pumps.ServiceBus/AzureServiceBusMessagePump.cs (hunk 20 above)
Comment (tomkerkhove) [reply src/Arcus.Messaging.Pumps.ServiceBus/AzureServiceBusMessagePump.cs]: Let's just follow the typical .NET approach by using `HasFlag`

**Comment by tomkerkhove** [comment]
https://github.com/arcus-azure/arcus.messaging/pull/59 should give you insights on the Docker container when tests fail

File: src/Arcus.Messaging.Pumps.ServiceBus/AzureServiceBusMessagePumpOptions.cs (hunk 16 above)
Comment (stijnmoreels) [reply src/Arcus.Messaging.Pumps.ServiceBus/AzureServiceBusMessagePumpOptions.cs]: Yeah, is OK, I just alwasy think in a "clean after you're done"-approach 🗑

**Comment by stijnmoreels** [comment]
> #59 should give you insights on the Docker container when tests fail

I think it's about the conenction string of the Service Bus topic? And that we have to set it to 'Manage'? 😝 But I find it strange, cause there's not a single shared access policy configured on the Service Bus Topic.... What is it using now, then? I wonder.

File: src/Arcus.Messaging.Pumps.ServiceBus/AzureServiceBusMessagePumpOptions.cs (hunk 16 above)
Comment (tomkerkhove) [reply src/Arcus.Messaging.Pumps.ServiceBus/AzureServiceBusMessagePumpOptions.cs]: Let's do that indeed, better

**Comment by tomkerkhove** [comment]
Looks like it's having issues with creating the subscription:
>  ---> System.ArgumentException: The specified HTTP verb (PUT) is not valid. To know more visit https://aka.ms/sbResourceMgrExceptions. . TrackingId:3c110d46-afef-4833-8574-f20f5834fd2c_G20, SystemTracker:arcus-messaging-dev-we-integration-tests:Queue:orders, Timestamp:2020-02-10T12:31:35
   --- End of inner exception stack trace ---
   at Microsoft.Azure.ServiceBus.Management.ManagementClient.SendHttpRequest(HttpRequestMessage request, CancellationToken cancellationToken)
   at Microsoft.Azure.ServiceBus.Management.ManagementClient.PutEntity(String path, String requestBody, Boolean isUpdate, String forwardTo, String fwdDeadLetterTo, CancellationToken cancellationToken)
   at Microsoft.Azure.ServiceBus.Management.ManagementClient.CreateSubscriptionAsync(SubscriptionDescription subscriptionDescription, RuleDescription defaultRule, CancellationToken cancellationToken)

File: src/Arcus.Messaging.Pumps.ServiceBus/AzureServiceBusMessagePump.cs [hunk 21]
```diff
@@ -49,6 +52,55 @@ public abstract class AzureServiceBusMessagePump<TMessage> : MessagePump<TMessag
...
+                JobId, SubscriptionName, serviceBusConnectionString.EntityPath);
+            
+            var subscriptionDescription = new SubscriptionDescription(serviceBusConnectionString.EntityPath, SubscriptionName)
+            {
+                UserMetadata = $"Subscription created by Arcus job: '{JobId}' to process Service Bus messages."
+            };
+
+            var ruleDescription = new RuleDescription("Accept-All", new TrueFilter());
+            var serviceBusClient = new ManagementClient(serviceBusConnectionString);
+            await serviceBusClient.CreateSubscriptionAsync(subscriptionDescription, ruleDescription, cancellationToken)
```
Comment (tomkerkhove) [inline src/Arcus.Messaging.Pumps.ServiceBus/AzureServiceBusMessagePump.cs]: YOu shuld check if the subscription with same name already exists

Comment (stijnmoreels) [reply src/Arcus.Messaging.Pumps.ServiceBus/AzureServiceBusMessagePump.cs]: Yep done. Thanks.

**Comment by stijnmoreels** [comment]
@tomkerkhove Ok, all fixed, the problem was that their was that the job ID was set to the subscription name in a Queue message pump when in that case the subscription name should be `null`.

File: src/Arcus.Messaging.Pumps.ServiceBus/AzureServiceBusMessagePump.cs [hunk 22]
```diff
@@ -49,6 +60,66 @@ public abstract class AzureServiceBusMessagePump<TMessage> : MessagePump<TMessag
...
+                UserMetadata = $"Subscription created by Arcus job: '{JobId}' to process Service Bus messages."
+            };
+
+            var ruleDescription = new RuleDescription("Accept-All", new TrueFilter());
+            var serviceBusClient = new ManagementClient(serviceBusConnectionString);
+
+            bool subscriptionExists = 
+                await serviceBusClient.SubscriptionExistsAsync(
+                    serviceBusConnectionString.EntityPath, SubscriptionName, cancellationToken);
+            if (subscriptionExists)
```
Comment (tomkerkhove) [inline src/Arcus.Messaging.Pumps.ServiceBus/AzureServiceBusMessagePump.cs]: Let's do this before we even start composing the rule description, log it as you have it now and return. No need for an else then

File: src/Arcus.Messaging.Pumps.ServiceBus/AzureServiceBusMessagePumpOptions.cs [hunk 23]
```diff
@@ -34,9 +34,28 @@ public class AzureServiceBusMessagePumpOptions
...
+
         /// <summary>
         ///     Default settings
         /// </summary>
-        internal static AzureServiceBusMessagePumpOptions Default => new AzureServiceBusMessagePumpOptions { AutoComplete = true };
+        internal static AzureServiceBusMessagePumpOptions Default => new AzureServiceBusMessagePumpOptions
+        {
+            AutoComplete = true,
+            JobId = Guid.NewGuid().ToString(),
+            TopicSubscription = TopicSubscription.CreateOnStart
```
Comment (tomkerkhove) [inline src/Arcus.Messaging.Pumps.ServiceBus/AzureServiceBusMessagePumpOptions.cs]: Didn't we decide to clean up after finished like you've suggested? If so, please don't forget to update docs with new default

Comment (stijnmoreels) [reply src/Arcus.Messaging.Pumps.ServiceBus/AzureServiceBusMessagePumpOptions.cs]: I suggested it, but wasn't sure you agreed 😄 .

File: src/Arcus.Messaging.Pumps.ServiceBus/AzureServiceBusMessagePump.cs (hunk 22 above)
Comment (stijnmoreels) [reply src/Arcus.Messaging.Pumps.ServiceBus/AzureServiceBusMessagePump.cs]: I'll move the rule description, use a try/finaly block for the service bus client to be closed, but I don't agree that removing the else and adding a return statement will benifit readability. If the return statement is forgotten, or removed by mistake, we got invalid code, while with the else this is much clearer.

File: src/Arcus.Messaging.Pumps.ServiceBus/AzureServiceBusMessagePumpOptions.cs (hunk 23 above)
Comment (tomkerkhove) [reply src/Arcus.Messaging.Pumps.ServiceBus/AzureServiceBusMessagePumpOptions.cs]: It's a fair suggestion so go ahead! 👍

File: docs/features/message-pumps/service-bus.md [hunk 24]
```diff
@@ -106,8 +106,8 @@ public void ConfigureServices(IServiceCollection services)
             options.JobId = Guid.NewGuid().ToString();
 
             // Indicate whether or not a new Azure Service Bus Topic subscription should be created/deleted
-            // when the message pump starts/stops (default: CreateOnStart).
-            options.TopicSubscription = TopicSubscription.CreateOnStart | TopicSubscription.DeleteOnStop;
+            // when the message pump starts/stops (default: DeleteOnStop, which implicitly means also 'create on start').
+            options.TopicSubscription = TopicSubscription.DeleteOnStop;
```
Comment (tomkerkhove) [inline docs/features/message-pumps/service-bus.md]: Ha think there is some miscommunication, I'd keep both by default.
```suggestion
            options.TopicSubscription = TopicSubscription.CreateOnStart | TopicSubscription.DeleteOnStop;
```

File: src/Arcus.Messaging.Pumps.ServiceBus/AzureServiceBusMessagePumpOptions.cs [hunk 25]
```diff
@@ -55,7 +55,7 @@ public class AzureServiceBusMessagePumpOptions
         {
             AutoComplete = true,
             JobId = Guid.NewGuid().ToString(),
-            TopicSubscription = TopicSubscription.CreateOnStart
+            TopicSubscription = TopicSubscription.DeleteOnStop
```
Comment (tomkerkhove) [inline src/Arcus.Messaging.Pumps.ServiceBus/AzureServiceBusMessagePumpOptions.cs]:
```suggestion
            TopicSubscription = TopicSubscription.CreateOnStart | TopicSubscription.DeleteOnStop;
```

File: docs/features/message-pumps/service-bus.md (hunk 24 above)
Comment (stijnmoreels) [reply docs/features/message-pumps/service-bus.md]: Yes, I know 😄 . `CreateOnStart = 1` and `DeleteOnStop = 3` so bitwise, if you choose `DeleteOnStop` you implicitly say that you want to create it on start. So it's the same.

Comment (stijnmoreels) [reply docs/features/message-pumps/service-bus.md]: But in the docs, we could make explicit indeed. In code, not very much added value I think.

Comment (tomkerkhove) [reply docs/features/message-pumps/service-bus.md]: Frankly, it should not be that way, how would I then delete it on stop without creating it here if I want?

Comment (stijnmoreels) [reply docs/features/message-pumps/service-bus.md]: It's implicit saying that you will create it at start. It's actually kind of a failsafe. So, that you can't configure the pump with only deleting the subscription, without also implicitly stating that you want it to be created at startup.
For readability, I can understand that it makes sense to use them both, but behind the scenes, there's no difference.

Comment (tomkerkhove) [reply docs/features/message-pumps/service-bus.md]: I don't like hidding things, we should just use a flag approach and if only `DeleteOnStop` is specified only do that. If the subscription is not there, throw an exception, it's their fault since they've configured it that way 🤷‍♂️

But default should be `CreateOnStart | DeleteOnStop` indeed.

Comment (tomkerkhove) [reply docs/features/message-pumps/service-bus.md]: Reason is - Sometimes subscriptions are created externally so we should not implicitly do this if it's not specified.

File: docs/features/message-pumps/service-bus.md [hunk 26]
```diff
@@ -57,25 +57,64 @@ public void ConfigureServices(IServiceCollection services)
...
+            // The amount of concurrent calls to process messages 
+            // (default: null, leading to the defaults of the Azure Service Bus SDK message handler options).
+            options.MaxConcurrentCalls = 5;
+
+            // The unique identifier for this background job to distinguish 
+            // this job instance in a multi-instance deployment (default: guid).
+            options.JobId = Guid.NewGuid().ToString();
+
+            // Indicate whether or not a new Azure Service Bus Topic subscription should be created/deleted
+            // when the message pump starts/stops (default: DeleteOnStop, which implicitly means also 'create on start').
```
Comment (tomkerkhove) [inline docs/features/message-pumps/service-bus.md]:
```suggestion
            // when the message pump starts/stops (default: CreateOnStart & DeleteOnStop).
```

**Comment by tomkerkhove** [comment]
Thanks!
